thrust/thrust/detail/vector_base.inl

```diff
@@ -879,105 +879,110 @@ void vector_base<T, Alloc>::append(size_type n)
-        // this copy overlaps
-        const size_type copy_length = (old_end - n) - position;
-        thrust::detail::overlapped_copy(position, old_end - n, old_end - copy_length);
+    // allocate exponentially larger new storage
+    new_capacity = ::cuda::std::max<size_type>(new_capacity, 2 * capacity());
 
-        // finally, fill the range to the insertion point
-        thrust::fill_n(position, n, x);
-      } // end if
-      else
-      {
-        // construct new elements at the end of the vector
-        m_storage.uninitialized_fill_n(end(), n - num_displaced_elements, x);
+    // do not exceed maximum storage
+    new_capacity = ::cuda::std::min<size_type>(new_capacity, max_size());
 
-        // extend the size
-        m_size += n - num_displaced_elements;
+    if (new_capacity > max_size())
+    {
+      throw std::length_error("insert(): insertion exceeds max_size().");
+    } // end if
 
-        // construct copy the displaced elements
-        m_storage.uninitialized_copy(position, old_end, end());
+    storage_type new_storage(copy_allocator_t(), m_storage, new_capacity);
 
-        // extend the size
-        m_size += num_displaced_elements;
+    iterator new_end = new_storage.begin();
 
-        // fill to elements which already existed
-        thrust::fill(position, old_end, x);
-      } // end else
-    } // end if
-    else
+    try
     {
-      const size_type old_size = size();
+      // Copy elements before insertion point
+      new_end = m_storage.uninitialized_copy(begin(), position, new_storage.begin());
 
-      // compute the new capacity after the allocation
-      size_type new_capacity = old_size + ::cuda::std::max THRUST_PREVENT_MACRO_SUBSTITUTION(old_size, n);
+      // Fill n elements with x at insertion point
+      new_storage.uninitialized_fill_n(new_end, n, x);
+      new_end += n;
 
-      // allocate exponentially larger new storage
-      new_capacity = ::cuda::std::max<size_type>(new_capacity, 2 * capacity());
+      // Copy elements after insertion point
+      new_end = m_storage.uninitialized_copy(position, end(), new_end);
+    } // end try
+    catch (...)
+    {
+      // something went wrong, so destroy & deallocate the new storage
+      new_storage.destroy(new_storage.begin(), new_end);
+      new_storage.deallocate();
 
-      // do not exceed maximum storage
-      new_capacity = ::cuda::std::min<size_type>(new_capacity, max_size());
+      // rethrow
+      throw;
+    } // end catch
 
-      if (new_capacity > max_size())
-      {
-        throw std::length_error("insert(): insertion exceeds max_size().");
-      } // end if
+    // call destructors on the elements in the old storage
+    m_storage.destroy(begin(), end());
 
-      storage_type new_storage(copy_allocator_t(), m_storage, new_capacity);
+    // record the vector's new state
+    m_storage.swap(new_storage);
+    m_size = old_size + n;
+  }
+  else
+  {
```

Comment:
Nitpick: I would prefer if we would return here and reduce indentation afterwards

```suggestion
    m_size = old_size + n;
    return;
  }

```